contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,495 @@
+  - [Open questions](#open-questions)
+- [Alternatives considered](#alternatives-considered)
+  - [Pods point at VPA](#pods-point-at-vpa)
+  - [VPA points at Deployment](#vpa-points-at-deployment)
+  - [Actuation using the Deployment update mechanism](#actuation-using-the-deployment-update-mechanism)
+  
+------------
+Introduction
+------------
+
+### Background ###
+* [Compute resources](http://kubernetes.io/docs/user-guide/compute-resources/)
+* [Resource QoS](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-qos.md)
+* [Admission Controllers](http://kubernetes.io/docs/admin/admission-controllers/)
+
+### Purpose ###
+Vertical scaling has two objectives:
+
+1. Reducing the maintenance cost, by automating configuration of resource
+requirements.
+
+2. Improving utilization of cluster resources, while minimizing the risk of containers running out of memory or getting CPU starved.
+
+### Related features ###
+#### Initial resources ####
+["Initial Resources"](https://github.com/kgrygiel/community/blob/master/contributors/design-proposals/initial-resources.md)
+is a very preliminary, proof-of-concept feature providing initial request based
+on historical utilization. It is designed to only kick in on Pod creation.
+VPA is intended to supersede this feature.
+
+#### In-place updates ####
+In-place Pod updates ([#5774]
+(https://github.com/kubernetes/kubernetes/issues/5774)) is a planned feature to
+allow changing resources (request/limit) of existing containers without killing them, assuming sufficient free resources available on the node.
+Vertical Pod Autoscaler will greatly benefit from this ability, however it is
+not considered a blocker for the MVP.
+#### Resource estimation ####
+Resource estimation is another planned feature, meant to improve node resource
+utilization by temporarily reclaiming unused resources of running containers.
+It is different from Vertical Autoscaling in that it operates on a shorter
+timeframe (using only local, short-term history), re-offers resources at a
+lower quality, and does not provide initial resource predictions.
+VPA and resource estimation are complementary. Details will follow once
+Resource Estimation is designed.
+
+------------
+Requirements
+------------
+
+### Functional ###
+
+1. VPA is capable of setting container resources (CPU & memory request/limit) at
+   Pod submission time.
+   
+2. VPA is capable of adjusting container resources of existing Pods, in
+   particular reacting to CPU starvation and container OOM events.
+   
+3. When VPA restarts Pods it respects the disruption budget.
+
+4. It is possible for the user to configure VPA with fixed constraints on
+   resources, specifically: min & max request.
+   
+5. VPA is compatible with Pod controllers, at least with Deployments.
+   In particular:
+     * Updates of resources do not interfere/conflict with spec updates.
+     * It is possible to do a rolling update of the VPA policy on an existing
```

Comment:
This point ("rolling update of the VPA policy") is about user updates to the *policy* (e.g. min resources or other knobs), not about updates of the actual resources initiated by VPA.
We want to assure that user change to the policy (e.g. increasing min resources) won't trigger a massive, simultaneous update of all affected pods.